Update docs for ByRefLike with generics for work in .NET 10

AaronRobinsonMSFT opened the pull request:
Revisit the IL sequence recognition approach and propose a compiler helper approach.

Add a few more tests for the current behavior of the runtimes.

Contributes to https://github.com/dotnet/runtime/issues/65112

AaronRobinsonMSFT committed:
Update docs for byreflike with generics

AaronRobinsonMSFT commented:
/cc @vzarytovskii @fanyang-mono

jkotas commented on docs/design/features/byreflike-generics.md:
```diff
@@ -15,7 +15,7 @@ Supporting ByRefLike type as Generic parameters will impact the following IL ins
 - `constrained.callvirt` &ndash; If this IL sequence resolves to a method implemented on `object` or default interface method, an error will occur during the attempt to box the instance.
 
-If any of the above instructions are attempted to be used with a ByRefLike type, the runtime will throw an `InvalidProgramException`. Sequences involving some of the above instructions are considered optimizations and represent cases that will remain valid regardless of a `T` being ByRefLike. See "Special IL Sequences" section below for details.
+If any of the above instructions are attempted to be used with a ByRefLike type, the runtime will throw an `InvalidProgramException`
```
```suggestion
If any of the above instructions are attempted to be used with a ByRefLike type, the runtime will throw an `InvalidProgramException`.
```

jkotas commented on docs/design/features/byreflike-generics.md:
```diff
@@ -24,6 +24,8 @@ The following instructions are already set up to support this feature since thei
 - `isinst` &ndash; Will always place `null` on stack.
 - `castclass` &ndash; Will always throw `InvalidCastException`.
 
+**NOTE** There are sequences involving some of the above instructions that are considered optimizations. These sequences represent cases that will remain valid regardless of a `T` being ByRefLike&mdash;see ["Special IL Sequences" section](#special_il_sequences) below for details.
```
```suggestion
**NOTE** There are sequences involving some of the above instructions that will remain valid regardless of a `T` being ByRefLike&mdash;see ["Special IL Sequences" section](#special_il_sequences) below for details.
```
I would not call these special sequences "optimizations". I see optimization as an optional operation that may or may not happen. It is not the case here. The special sequences must be recognized for the program to work.

AaronRobinsonMSFT replied:
Agree.

AaronRobinsonMSFT committed:
Apply suggestions from code review

Co-authored-by: Jan Kotas <jkotas@microsoft.com>

AaronRobinsonMSFT commented on docs/design/specs/Ecma-335-Augments.md:
```diff
@@ -1026,6 +1027,27 @@ Changes to signatures:
+New sub-section should be added after III.4.33 that describes sequences of IL instructions that can be used on ByRefLike types when using within a generic context.
+
+#### III.4.X
+The following are IL sequences involving the `box` instruction. They are used for on ByRefLike types and shall be valid in cases where the result can be computed at run-time and elided safely, through JIT compilation or interpretation. These sequences must now be elided when the target type is ByRefLike. The conditions where each sequence is elided are described below.
```
```suggestion
The following are IL sequences involving the `box` instruction. They can be used on ByRefLike types and shall be valid in cases where the result can be computed at run-time and elided safely&mdash;through JIT compilation or interpretation. These sequences must now be elided when the target type is ByRefLike. The conditions where each sequence is elided are described below.
```

AaronRobinsonMSFT committed:
Apply suggestions from code review

jkotas commented on docs/design/specs/Ecma-335-Augments.md:
```diff
@@ -1026,6 +1027,27 @@ Changes to signatures:
+
+### II.10.1.7
+An additional IL keyword, `byreflike`, is introduced to indicate use of ByRefLike types is permitted. This expands the set of permissible types used by this parameters, but limits the potential instructions that can be used on instances of this generic parameter type.
+
```
Also, II.23.1.7 - add name and encoding of the new bit

jkotas commented on docs/design/specs/Ecma-335-Augments.md:
```diff
@@ -1026,6 +1027,27 @@ Changes to signatures:
+An additional IL keyword, `byreflike`, is introduced to indicate use of ByRefLike types is permitted. This expands the set of permissible types used by this parameters, but limits the potential instructions that can be used on instances of this generic parameter type.
+
+### III.4
+New sub-section should be added after III.4.33 that describes sequences of IL instructions that can be used on ByRefLike types when using within a generic context.
```
Link to the spec or include the list of sequences inline here.

AaronRobinsonMSFT replied:
That is right below in `III.4.X`

jkotas replied:
Ah ok, I have missed that the next paraph talks about the same section.

AaronRobinsonMSFT committed:
Feedback.

AaronRobinsonMSFT committed:
Offline feedback.

AaronRobinsonMSFT committed:
Add additional tests based on feedback.

AaronRobinsonMSFT committed:
Add check at run-time to CoreCLR and native AOT.
This is needed when generics parameters are involved.

AaronRobinsonMSFT committed:
New JIT helper mean rev'ing JIT interface ID.

AaronRobinsonMSFT committed:
Merge remote-tracking branch 'upstream/main' into update_ecma335

AaronRobinsonMSFT committed:
Update JIT interface GUID for new JIT helper.

AaronRobinsonMSFT committed:
Apply JIT format

AaronRobinsonMSFT commented:
@jakobbotsch Can you check out the updates to the importer using the new JIT helper?

AaronRobinsonMSFT commented:
@jaredpar @AlekseyTs @agocke I've updated the docs and ECMA description regarding the IL instruction sequences. I've tried to be as clear as possible with them. Please let me know if these are sufficient.

@jkotas @davidwrighton @MichalStrehovsky I added a new JIT helper for checking at run-time if a ByRefLike type can be cast to another type. This is purely about checking the type, not if one can call it. Note that calling a default interface method is still an issue in this case and not one that can be solved in the current PR.

@lambdageek @fanyang-mono I believe mono's implementation is already valid for the new tests and no changes should be needed. I do ask that you refer to the latest language in the spec.

@stephentoub I ask that you stop being curious... whenever you are, I need to do more work :-)

jkotas commented on docs/design/specs/Ecma-335-Augments.md:
```diff
@@ -1026,6 +1027,36 @@ Changes to signatures:
+
+`box` ; `isinst` ; `unbox.any` &ndash; Becomes a NOP, if the box, `isinst`, and unbox target types are all equal.
+
+`box` ; `isinst` ; `br_true/false` &ndash; Becomes a constant, if the box target type is ByRefLike or the box target type is `Nullable<T>` and target type equalities are computed to be equal. The sequence will also be elided if the box target type is a ByRefLike type, but needs to be checked at run-time, not JIT compile time. This latter case is common when the box target type uses Generic parame ...
```
Is it valid to say that all these special sequences produce the exact same result as if the byref-likeness of the types involved was dropped? If it is the case, I think it would be easier way to explain the behavior. If it is not the case, this is getting very complicated.

AaronRobinsonMSFT replied:
I tried to update it with similar language. Please take a look.

jkotas commented:
> Note that calling a default interface method is still an issue in this case and not one that can be solved in the current PR.

What is this problem about?

AaronRobinsonMSFT commented on src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/CastHelpers.cs:
```diff
@@ -316,6 +319,10 @@ internal static unsafe class CastHelpers
         }
 
+        [DebuggerHidden]
+        private static int ChkCastTypeToType(void* fromTypeHnd, void* toTypeHnd)
```
I'm just going with what the JIT is expecting. In `importer.cpp`, I am inserting code with a return type of `TYP_INT` so am following that pattern. Is returning managed `bool` here appropriate?

AaronRobinsonMSFT replied:
`IsAssignable` makes sense too. I will take that suggestion.

jkotas replied:
> Is returning managed bool here appropriate?

It would look better. (There should not be any low-level difference since bool is physically returned as INT32 in managed calling convention.)

AaronRobinsonMSFT replied:
Okay.

AaronRobinsonMSFT committed:
Feedback on rename

AaronRobinsonMSFT committed:
Merge remote-tracking branch 'upstream/main' into update_ecma335

AaronRobinsonMSFT committed:
Update JIT interface ID

jkotas commented on src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/CastHelpers.cs:
```diff
@@ -316,6 +319,10 @@ internal static unsafe class CastHelpers
         }
 
+        [DebuggerHidden]
+        private static bool IsAssignable(void* fromTypeHnd, void* toTypeHnd)
```
- Should we use consistent name? Ether `IsAssignable` or `AreTypeAssignable` is fine me - as long it is used consistently.

- Should this follow the same pattern as other casting helper and do a lookup in cast-cache first before calling the slow "NoCache" path?

AaronRobinsonMSFT replied:
> Should we use consistent name?

We are consistent. The JIT helper is called `IsAssignable()` for both versions (CoreCLR and native AOT), and both call an `AreTypesAssignable()` method as their implementation.

> Should this follow the same pattern as other casting helper and do a lookup in cast-cache first before calling the slow "NoCache" path?

Currently, both native AOT and CoreCLR cache, but the caching is done within the underlying `AreTypesAssignable` implementations. For native AOT the cache can be observed in the same file in `AreTypesAssignableInternal`. For CoreCLR, the caching is done within the implementation of `TypeHandle::CanCastTo()`

jkotas replied:
> The JIT helper is called IsAssignable() for both versions (CoreCLR and native AOT), and both call an AreTypesAssignable() method as their implementation.

I would call the JIT helper `AreTypesAssignable` as well.

> For CoreCLR, the caching is done within the implementation of TypeHandle::CanCastTo()

This implementation is leaving performance on the table. All existing managed casting helpers in this file are doing the cache lookup on the managed side to avoid paying managed/unmanaged transition costs for cache hit, and only call the NoCacheLookup path with managed/unmanaged transition path on cache miss.

AaronRobinsonMSFT committed:
Merge remote-tracking branch 'upstream/main' into update_ecma335

MichalStrehovsky commented on src/coreclr/nativeaot/Runtime.Base/src/System/Runtime/TypeCast.cs:
```diff
@@ -475,6 +475,10 @@ private static unsafe object CheckCastClassSpecial(MethodTable* pTargetType, obj
             return ThrowInvalidCastException(pTargetType);
         }
 
+        [RuntimeExport("RhTypeCast_IsAssignable")]
+        public static unsafe bool IsAssignable(MethodTable* fromTypeMT, MethodTable* toTypeMT)
+            => AreTypesAssignable(fromTypeMT, toTypeMT);
+
```
The method this is calling into is already exposed as the `RhTypeCast_AreTypesAssignable` RuntimeExport. Do we need a new RuntimeExport? Does this API have the expected semantics? (`AreTypesAssignable` assumes a boxed source so if `fromTypeMT` is a struct and `toTypeMT` is System.Object or System.ValueType, it will return true.)

AaronRobinsonMSFT replied:
The semantics here appear correct. I've added tests for the boxing source semantics.

jakobbotsch commented on src/coreclr/jit/importer.cpp:
```diff
@@ -3021,6 +3021,25 @@ int Compiler::impBoxPatternMatch(CORINFO_RESOLVED_TOKEN* pResolvedToken,
                                 return returnToken;
                             }
+
+                            // Casts that result to TypeCompareState::May and involve a ByRefLike type,
+                            // must be checked at run-time.
+                            if (opts == BoxPatterns::IsByRefLike)
+                            {
+                                assert(castResult == TypeCompareState::May);
+                                JITDUMP(
+                                    "\n Importing BOX; ISINST; BR_TRUE/FALSE for ByRefLike, must be checked at run-time\n");
+
+                                impSpillSideEffects(false, CHECK_SPILL_ALL DEBUGARG("spilling side-effects"));
+                                impPopStack();
+
+                                GenTree* toType   = impTokenToHandle(pResolvedToken);
+                                GenTree* fromType = impTokenToHandle(&isInstResolvedToken);
+                                impPushOnStack(gtNewHelperCallNode(CORINFO_HELP_ISASSIGNABLE, TYP_INT, toType,
+                                                                   fromType),
+                                               typeInfo(TYP_INT));
+                                return returnToken;
+                            }
```
Does the behavior of this helper differ from the fallback case where we did not recognize the pattern here and just ended up in the standard `isinst` importation path?

AaronRobinsonMSFT replied:
Yes, in the sense that it is narrower and doesn't result in a deoptimization in `impCastClassOrIsInstToTree`. Instead the result is practically identical to the optimization path, but even narrower because it is exclusively for the `BoxPatterns::IsByRefLike` case.

jakobbotsch replied:
Gotcha, I think I understand now.

AaronRobinsonMSFT committed:
Review feedback.

AaronRobinsonMSFT committed:
Update spec.

agocke commented on docs/design/features/byreflike-generics.md:
```diff
@@ -112,21 +114,21 @@ Adding `gpAcceptByRefLike` to the metadata of a Generic parameter will be consid
+`box` ; `br_true/false` &ndash; Becomes the constant `true`, if the box target type is non-`Nullable<T>`.
 
-`box` ; `isinst` ; `unbox.any` &ndash; The box, `isint`, and unbox target types are all equal.
+`box` ; `isinst` ; `unbox.any` &ndash; Becomes a NOP, if the box, `isinst`, and unbox target types are all equal.
```
I don't understand how these sequences affect the stack state. For example, if Roslyn does

```C#
void M<T>(T t) {
  if (t is Span<int> s) {
}
}
```

That's gonna be a `box; isinst; unbox.any`, but the key thing is that the stack started with a `T` on top, and now it should have `Span<int>`. Otherwise you couldn't legally do a `stloc` to a `Span<int>` local.

AaronRobinsonMSFT replied:
It is a NOP, the pushed state remains. Yes, it will remain `T`.

agocke replied:
In that case it's not clear to me why we're doing this. What scenario does this enable?

AaronRobinsonMSFT replied:
This was an existing recognized sequence that is optimized, no new. It isn't unreasonable if we'd like to skip documenting it at this time.

jakobbotsch reviewed (approved):
JIT part LGTM

jaredpar commented on docs/design/features/byreflike-generics.md:
```diff
@@ -248,4 +250,13 @@ class B : A
+class A
+{
+    public bool M<T, U>(T t) where T: allows ref struct
+        => t is U;
```
This sample does not have `t is U u`. Is that intentional? Essentially, that the code gen today only supports type testing not storing into to type?

AlekseyTs commented on docs/design/features/byreflike-generics.md:
```diff
@@ -112,21 +114,21 @@ Adding `gpAcceptByRefLike` to the metadata of a Generic parameter will be consid
+`box` ; `unbox.any` &ndash; Becomes a NOP, if the box target type is equal to the unboxed target type.
 
-`box` ; `br_true/false` &ndash; The box target type is non-`Nullable<T>`.
+`box` ; `br_true/false` &ndash; Becomes the constant `true`, if the box target type is non-`Nullable<T>`.
```
This sounds somewhat confusing. The sequence is effectively a conditional branch, it doesn't produce a value and, therefore, cannot become a constant. I am not sure if it is really important whether the sequence can be elided or not. What, I think, is important is that the sequence remains valid for all target types and can be interpreted as behaving as if the ByRefLike nature of the target type doesn't exist. If that is accurate, consider adjusting description accordingly

AlekseyTs commented on docs/design/features/byreflike-generics.md:
```diff
@@ -112,21 +114,21 @@ Adding `gpAcceptByRefLike` to the metadata of a Generic parameter will be consid
+`box` ; `isinst` ; `unbox.any` &ndash; Becomes a NOP, if the box, `isinst`, and unbox target types are all equal.
 
-`box` ; `isinst` ; `br_true/false` &ndash; The box target type is equal to the unboxed target type or the box target type is `Nullable<T>` and target type equalities can be computed.
+`box` ; `isinst` ; `br_true/false` &ndash; Can become a constant, if the box target type is ByRefLike or the box target type is `Nullable<T>` and target type equalities are computed to be equal. This sequence can be interpreted as behaving as if the ByRefLike nature of the input type doesn't exist.
```
Similar comment here. I think that for IL writers it is not that important whether the sequence can be elided. What is important is that the sequence remains valid for any combination of types and this sequence can be interpreted as behaving as if the ByRefLike nature of the types involved doesn't exist. If that is accurate, consider saying just that.

AaronRobinsonMSFT committed:
Updates

jaredpar commented on docs/design/features/byreflike-generics.md:
````diff
@@ -114,21 +114,103 @@ Adding `gpAcceptByRefLike` to the metadata of a Generic parameter will be consid
+    if (t is U u)          // Invalid
+    if (t is Span<U> su)   // Invalid
+}
+```
````
Can you help us understand a bit how this behaves with actual types. Consider as an example: 

```csharp
// Not real C#, just using as an example to understand the IL semantics
ref struct RS2 : I { }

// Which of the `if` succeed and which fail here?
M<RS2, I>(); 
```

I'm particularly interested in the following ones:

```csharp
    if (t is I itf) 
    if (t is object)   // I added this one
    if (t is U u)   
```

Sorry if this seems odd but knowing how IL would execute these helps us understand how the C# rules need to be constructed. Consider concretely how C# works today with non-generic code

```csharp
ref struct S : I { }
interface I { }
void Example(S s) {
  // Warning: expression always false
  if (s is I) 
}

// This is what the code would look like if the user decided to make it generic. 
void GenericExample<T>(T t ) where T: allows ref struct {
  // What do we do here?
  if (t is I) 
}
```

Knowing how that will execute in IL really helps us nail down what we want to recommend to C# LDM on what the C# rules will be. Think we already have a fairly good understanding of what we want to recommend but really want to _know_ the IL semantics to ensure the design is right.

AaronRobinsonMSFT replied:
Given the call `M<RS2, I>()`:

```csharp
    if (t is I itf) 
```
Can be made to work, but the type local would need to be `RS2` since it cannot be boxed, so perhaps this is "Invalid" from the C# perspective. The check is safe to make (that is, `if (t is I)`).

```csharp
    if (t is object)
```
The `object` check will evaluate to `true` in this PR. It is unlikely we would change that behavior if we adopted the IL instruction approach. We could update the helper functions to special case ByRefLike behavior or define in a special manner to be used when the `allows ref struct` constraint is added.

```csharp
    if (t is U u)
```
This will practically always result in a `false` condition for obvious reasons. If `U` were constrained to something it could potential be the same as `if (t is I itf)`.

Going down the IL instruction approach we need more precision, but adopting the Compiler helpers means we can permit more. The helpers could be "general purpose" in the sense they entirely replace the aforementioned sequences or we could make the helpers ByRefLike only and for example evaluate `t is object` to `false`.

I don't think anyone on the runtime team has a strong opinion about the Compiler helpers route, but if we continue down the IL sequence path, I think we need to know C# wants and then we can say what is possible to guarantee.

```csharp
// This is what the code would look like if the user decided to make it generic. 
void GenericExample<T>(T t ) where T: allows ref struct {
  // What do we do here?
  if (t is I) 
}
```

The "check", `t is I`, is always possible based on this PR and the proposed Compiler helper would provide the same behavior at present. The question though becomes, what would one want to do in that branch if it evaluated to `true`?

jaredpar replied:
> The object check will evaluate to true in this PR.
> The "check", t is I, is always possible based on this PR and the proposed Compiler helper would provide the same behavior at present

I understand why this is sensible from an IL / CLR perspective. From a C# perspective though it's not in alignment with how ref safety and type checks work. Both of these types of checks evaluated to `false` at compile time with concrete types + a warning. Given this behavior from IL, in generic contexts we'd need to block type checks with potentially fall into these categories with errors. That is the only way to ensure there is a consistent experience between generic and non-generic code. 

That is not necessarily wrong for runtime or C#. It's nice when the C# and IL type system are in full agreement. But there are other cases where aren't . 

Note: in retrospect kinda wish we'd pushed for an error in the non-generic case.

jkotas replied:
If we go with helper methods, we can make them do whatever works best for C# in these corner cases.

AaronRobinsonMSFT replied:
> If we go with helper methods, we can make them do whatever works best for C# in these corner cases.

Agreed. This concept was what made Jan's statement so appealing. We just need to define what makes sense for C# and then the really hard part, naming the APIs...

AaronRobinsonMSFT replied:
`bool ByRefLikeAwareIsInstanceOf<TFrom, TTo>(TFrom source)` and `TTo ByRefLikeAwareCastTo<TFrom, TTo>(TFrom source)`?

AaronRobinsonMSFT committed:
Updates

AaronRobinsonMSFT committed:
Updates

jkotas commented on docs/design/features/byreflike-generics.md:
```diff
@@ -13,16 +13,18 @@ Supporting ByRefLike type as Generic parameters will impact the following IL ins
+- `constrained.callvirt` &ndash; When applied to a ByRefLike instance, if this IL sequence resolves to a method implemented on `object` or a default interface method, an error will occur during the attempt to box the ByRefLike instance.
 
-If any of the above instructions are attempted to be used with a ByRefLike type, the runtime will throw an `InvalidProgramException`. Sequences involving some of the above instructions are considered optimizations and represent cases that will remain valid regardless of a `T` being ByRefLike. See "Special IL Sequences" section below for details.
+If any of the above instructions are attempted to be used with a ByRefLike type, the runtime will throw an `InvalidProgramException`.
```
This is not accurate (in the current implementation at least). For example, `newarr` used with ByRefLike type throws:

`Unhandled exception. System.TypeLoadException: Could not create array type 'System.Span`1[System.Int32][]' from assembly 'System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e' because the element type is ByRef-like.`

jkotas commented on docs/design/features/byreflike-generics.md:
```diff
@@ -13,16 +13,18 @@ Supporting ByRefLike type as Generic parameters will impact the following IL ins
 - `newarr` / `stelem` / `ldelem` / `ldelema` &ndash; Arrays are not able to contain ByRefLike types.
     - `newobj` &ndash; For multi-dimensional array construction.
-- `constrained.callvirt` &ndash; If this IL sequence resolves to a method implemented on `object` or default interface method, an error will occur during the attempt to box the instance.
+- `constrained.callvirt` &ndash; When applied to a ByRefLike instance, if this IL sequence resolves to a method implemented on `object` or a default interface method, an error will occur during the attempt to box the ByRefLike instance.
```
Should we be more specific about the error? What is the exception that this is going to throw?

AaronRobinsonMSFT replied:
Good point. I went through and updated this section. I also ensured we had tests for these scenarios.

AaronRobinsonMSFT committed:
Updates

jkotas commented on docs/design/features/byreflike-generics.md:
```diff
@@ -9,20 +9,23 @@ Using ByRefLike types in Generic parameters is possible by building upon support
 
 Supporting ByRefLike type as Generic parameters will impact the following IL instructions:
 
+Throws `InvalidProgramException`:
```
Should we be more explicit about the allowed timing of these exceptions? I assume that they can either thrown when method is JITed (ie the whole method is replaced with InvalidProgramException) or they can be thrown only once the specific IL instruction is executed. I assume that that the exact timing is unspecified - both options are compliant implementations. Is that correct?

The answer to this question influences where the following is a valid method with predictable behavior or not.
```
string M<T>(T value) where T: allows ref struct
{
     if (typeof(value).IsByRefLike) return "";

     return value.ToString(); // Assume that this is compiled as constrained.callvirt 
}
```

AaronRobinsonMSFT replied:
> Should we be more explicit about the allowed timing of these exceptions?

I'm not entirely sure in this case. In fact, I'm now struggling to understand what is too much details vs what what is too little. This seems at odds with some of the early feedback about when I was describing what the JIT was going to do (for example, elide instructions). Calling out the JIT or native AOT or "when" this occurs seems to be at odds with the abstraction of IL and getting into when the IL is evaluated or translated. Should I then also consider the behavior in the interpreter case?

jkotas replied:
I think it is important to define the valid IL usages with well-defined behavior and declare everything else as invalid IL with undefined behavior. We do not need to spec the undefined behavior. We like to throw
InvalidProgramException or TypeLoadException for the undefined behavior as a diagnostic aid, but we typically do not guarantee whether the exact exception that is going to be thrown and the undefined behavior can vary between runtimes variants and versions.

It is not clear to me whether these exceptions are part of the well-defined behavior space or undefined behavior space. For example, it is not clear to me whether the following has well-defined behavior:

```
string M<T>(T value) where T: allows ref struct
{
     if (typeof(value).IsByRefLike) return "";

     return value.ToString(); // Assume that this is compiled as constrained.callvirt 
}

M<string>("Hello");
```

Is this a valid program that is guaranteed to work or is this invalid program with undefined behavior?

What about `M<Span<int>>("Hello")`. Is that a valid program that is well-defined behavior or is it an invalid program with undefined behavior?

AaronRobinsonMSFT committed:
Feedback

jkotas commented on docs/design/features/byreflike-generics.md:
```diff
@@ -7,22 +7,25 @@ Using ByRefLike types in Generic parameters is possible by building upon support
-- `box` &ndash; Types with ByRefLike parameters used in fields cannot be boxed.
+When used with a ByRefLike type these instructions result in undefined behavior:
+- `box` &ndash; When applied to ByRefLike types used in fields cannot be boxed.
+- `constrained.callvirt` &ndash; If this IL sequence resolves to a method implemented on `object` or a default interface method.
```
@jaredpar Are we happy that you can produce programs with undefined behavior with a safe vanilla C#? Is there a prior art like that?

In the current implementation, this will typically produce `InvalidProgramException`. The docs for `InvalidProgramException` say "Generally this indicates a bug in the compiler that generated the program.", so people are likely going to filling bugs on this in the Roslyn repo.

AlekseyTs replied:
> Are we happy that you can produce programs with undefined behavior with a safe vanilla C#? Is there a prior art like that?

Regarding a prior art. I am not sure about the specific exception, but one could run into an exception trying to invoke a default interface implementation which is ambiguous. Compiler guards against this situation when it compiles a type implementing the interface, but things can change after the type is compiled.

Regarding this specific situation. Compiler disallows calling object methods on ByRefLike type parameters. Compiler disallows calling non-virtual interface instance methods on ByRefLike type parameters. Compiler also requires `ref` structs to implement all virtual members of an implemented interface. So, if all parts are compiled in C# and interface members aren't added after the fact, one shouldn't run into a situation when the exception is thrown.

jkotas replied:
> one could run into an exception trying to invoke a default interface implementation which is ambiguous

The exception in this case is not InvalidProgramException and the behavior is fully deterministic. The exception is always thrown by the call itself, and it won't cause the method to fail to execute in its entirely. It is not that different from other dynamic type system violations, like InvalidCastException from castclass that is also fully determistic.

> interface members aren't added after the fact

Adding interface members after the fact has been the main motivation for default interface methods.

I am starting to think we need to make these behaviors fully deterministic in this case too.

agocke replied:
I agree that the runtime behavior should be deterministic.

On the compiler side, I don't think there's any way the compiler could prevent the situation from happening. Default interface methods can appear after compilation. So this has to be a runtime check.

I would expect that invoking a default interface implementation on ref struct would deterministically produce some well-defined exception, ideally one that precisely explains the situation.

jaredpar replied:
> On the compiler side, I don't think there's any way the compiler could prevent the situation from happening.

Agree the compiler can't fully prevent the DIM situation. At the same time though for this to happen the user must within the same method declare: 

1. I want to allow `T` to be a `ref struct`
2. I want to call a DIM method on `T` 

That is not wrong but it is to some degree contradictory / inviting this problem to happen. It's enough of sketchy case that part of the design suggested we could even [warn in this case](https://github.com/dotnet/csharplang/blob/main/proposals/ref-struct-interfaces.md#warn-on-dim-invocation) if we desired (LDM did not decide to do this). 

As long as the runtime has deterministic behavior here, preferably with a message or type that is easy to search on, then I think we are okay because this is a pretty narrow case. It can be added to the existing documentation around DIM.

AaronRobinsonMSFT replied:
> I am starting to think we need to make these behaviors fully deterministic in this case too.
> I agree that the runtime behavior should be deterministic.

I would like to understand more about these phrases. There are a lot words and phrases being used and I'm really not following all of them. What does deterministic mean in this case? We have multiple modes of execution. Does it fail at JIT/compilation? Does it partially execute? Does it fail during interpretation? It seems given other comments that "undefined behavior" is defined as "may fail to execute partially or in its entirety", is this the definition for all uses of "undefined behavior"?

jkotas replied:
> What does deterministic mean in this case? 

It means that the following test passes, on all runtime variants (JIT, interpretter, AOT, CoreCLR, Mono, ...):

```csharp
Assert.Equal(1, M<S>(false));

Assert.Throws<TBDException>(() => M<S>(true));

int M<T>(T value, bool callInterfaceMethod) where T: allows ref struct
    => callInterfaceMethod ? value.M() : 0;

interface IMy
{
     int M() => 1;
}

ref struct S : IMy
{
}
```

`TBDException` can be either `NotSupportedException` or we can introduced a new dedicated exception for this situation (I am not sure whether it is worth it).

agocke replied:
Yup that is one possibility. I used the term “deterministic” but I really mean fully defined behavior. I will leave the exact semantics up to you as I don’t have a strong preference. However, my suggestion would be we narrowly scope the exception to the problematic location: the call instruction. 

Specifically, imagine the following was present at compile time:

```C#
M(new S());

void M<T>(T value) where T: I, allows ref struct
{
    value.M1();
    value.M2();
}

ref struct S : I
{
    void M1() {}
    void M2() {}
}
interface I
{
   void M1();
   void M2();
}
```

This is all valid C#. Now imagine that I is in another assembly and adds M3 with a default interface method. I think this program should continue working without throwing an exception. That means, I think the exception should be narrowly scoped to the invocation of a default interface method, not at type load. 

Similarly, imagine the very complex scenario where the top-level M, I, and S are all in different assemblies. I adds an M3 with a default implementation. The M method calls M3 after M2, but S is never recompiled so it never implements M3. I think the defined behavior should be that calls M1 and M2 succeed. It’s only at the call site to M3 where an exception is thrown. 

The reason I like this computational model is that I think it conceptually matches the problem at the user level: the problem with ref structs and default interface implementations is that the receiver requires boxing. Therefore I think the failure should occur at the place where the problem is: boxing the receiver. At the C# level this is nominally the call site, but for IL this could be at the location where we would otherwise box the receiver.

AaronRobinsonMSFT replied:
> ```csharp
> int M<T>(T value, bool callInterfaceMethod) where T: allows ref struct
>     => callInterfaceMethod ? value.M() : 0;
> ```

> That means, I think the exception should be narrowly scoped to the invocation of a default interface method, not at type load.

I think this is a bad argument and isn't something we should support. I think we're trying to impose runtime semantic behavior that shouldn't exist based on all the previous talk about not wanting to create invalid IL sequences and also about introducing a branching invariant that isn't related to the underlying issue - the DIM.

Based on the current ECMA-355, the above is invalid IL and hence `InvalidProgramException` is appropriate. We decided that we shouldn't change anything which is where this is coming from.

The second is the argument about the branch condition. I really dislike this because it is unlikely to represent the troublesome concept - the DIM. One can imagine a scenario where, using reflection, people handle this case, but the `callInterfaceMethod` predicate will never represent that issue. This is especially true in the example of about someone introducing a new DIM on a type.

I really don't see the practical utility here. With the compiler helpers proposal below, the above IL sequences would never to exposed for ByRefLike types. This means we can leave the description as before and it remains invalid IL when the type is ByRefLike and must be boxed.

jkotas replied:
> With the compiler helpers proposal below, the above IL sequences would never to exposed for ByRefLike types.

I do not think this is correct. The above IL sequence can be exposed to ByRefLike types through libraries evolving independently, and one of the libraries adding default interface method implementation.

`AmbiguousMatchException` is existing prior art for similar situation. The ambiguous match is also something that can be introduced by libraries evolving independently. We could have said that the ambiguous match is invalid IL and leave the behavior undefined. We have not done that. Instead, we have made the behavior well defined, and introduced a dedicated exception for it that is thrown at the callsite with the problem. A program with ambiguous match is valid IL.

AaronRobinsonMSFT replied:
> The above IL sequence can be exposed to ByRefLike types through libraries evolving independently, and one of the libraries adding default interface method implementation.

It isn't just adding a default inteface method implementation (DIM impl), it also requires adding a call to it. If there was some way that simply adding a DIM impl caused this to occur, I do get it, but disagree it is all that interesting to change behavior.

> We could have said that the ambiguous match is invalid IL and leave the behavior undefined. We have not done that.

That choice though would be adding to ECMA when something became invalid IL. In this case the definition of invalid IL remains the same, we are relaxing it and thus need to add to ECMA. That is the part I pushing back on. The definition as-is covers this case and we should accept that.

I do accept that if independently evolving libraries can introduce this issue with one not having to recompile then I may need to reconsider, but I don't see how that is presently a common enough scenario to be concerned with. There are three (3) steps that need to happen.

1) The `allows ref struct` constraint needs to be added to a type constrained Generic parameter.
2) The use of the type constrained Generic parameter from (1) needs to call a potential DIM impl on a ByRefLike type.
3) The ByRefLike type in (2) needs to implement an interface where a new DIM impl is added.

I agree that it is conceivable that the interface from (3) is updated, (1) is updated, and then issues occur. This seems very clear the causality though and there is nothing that a specific exception is going to clarify in this case. In my opinion, enabling the branch support here actually hides this case far more since the underlying issue DIM on ByRefLike type, the underlying issue, is being blocked by some unrelated type predicate.

I simply don't see how this makes anything safer and personally see this as creating an architectural time bomb for someone to hit later.

AaronRobinsonMSFT replied:
For context, the current message for `constrained.callvirt` when a DIM or `object` method is called is:

CoreCLR:
> `System.InvalidProgramException: Cannot create boxed ByRef-like values.`

native AOT:
> `System.InvalidProgramException: Common Language Runtime detected an invalid program. The body of method '<METHOD NAME>' is invalid.`

jkotas replied:
> I agree that it is conceivable that the interface from (3) is updated, (1) is updated, and then issues occur.

Right. It means that user can end up in undefined behavior by doing a sequence of compatible changes. If a sequence of compatible changes leads to an undefined behavior in a vanilla C#, it is a flaw in the design of the feature.

> an architectural time bomb for someone to hit later.

I do not see how this is an architectural timebomb. We just need to declare this behavior to well-defined in the spec and update the runtimes to behave according to the spec. Do you see that as particular hard to do?

AaronRobinsonMSFT replied:
> If a sequence of compatible changes leads to an undefined behavior in a vanilla C#, it is a flaw in the design of the feature.

My statement "issues occur" is about the unexpected throwing of `InvalidProgramException`. I think we're talking past one another and I accept this is because of what I've written about "undefined behavior". There is nothing undefined here. An `InvalidProgramException` is going to be thrown because the IL is invalid in this scenario - it is fully defined. This means we are remain safe here.

If we have agreement with the above, then (a) I need to remove the "undefined behavior" in the design document and (b) we can then have a discussion about whether we want a more specific exception or not.

For (a), I am going to be adding additional tests that allows us to confirm the defined behavior.

jkotas replied:
I would like to see the exact timing to be defined too. I think that the most sensible behavior is to throw at the callsite.

I do not think that it should throw InvalidProgramException. InvalidProgramException suggests a bug in the compiler (or a bug in unsafe code). It is not the case here.

AaronRobinsonMSFT replied:
> I would like to see the exact timing to be defined too.

I disagree, but it doesn't look like we are going to align on this. If I were to adopt this perspective, throwing at the callsite is reasonable.

> I do not think that it should throw InvalidProgramException. InvalidProgramException suggests a bug in the compiler (or a bug in unsafe code). It is not the case here.

Again, I disagree, but I'm not sure how to bridge this gap at present. I'll update this design to insert a placeholder.

jkotas replied:
Ok. Do you agree or disagree with the following points?

1. A sequence of compatible changes (compatible from the type system point of view) should not lead to an invalid program or undefined behavior in a pure C#.  
2. Ambiguous default interface method resolution is very similar situation as default interface method invocation on byref-like type.
3. Throwing at the callsite to handle ambiguous default interface method resolution has been the right design. (If we can do it again, we would do it the same way.)

These points are behind my though process.

AaronRobinsonMSFT replied:
**TL;DR** I see your point now. I agree with your preference here on both the documentation of time of failure and call-site behavior.

I agree with all those points. I was struggling with (1) for a bit, but that was a confusion on my part with the proposed scenario - playing SharpLap more helped. I asked around and now I get the issue mentioned in (2), it makes me sad. Point (3) is where the main mental hurdle remained; however, given my new understanding of (2), point (3) follows directly. I will update the design as suggested, appreciate the patience.

jkotas commented on docs/design/features/byreflike-generics.md:
```diff
@@ -7,22 +7,25 @@ Using ByRefLike types in Generic parameters is possible by building upon support
+Supporting ByRefLike type as Generic parameters will impact the following IL instructions.
 
-- `box` &ndash; Types with ByRefLike parameters used in fields cannot be boxed.
+When used with a ByRefLike type these instructions result in undefined behavior:
```
```suggestion
When used with a ByRefLike type these instructions result in undefined behavior (the method containing the IL instruction may fail to execute partially or in its entirety):
```

AaronRobinsonMSFT committed:
Update docs/design/features/byreflike-generics.md

Co-authored-by: Jan Kotas <jkotas@microsoft.com>

AaronRobinsonMSFT committed:
Update tests

AaronRobinsonMSFT committed:
Merge remote-tracking branch 'upstream/main' into update_ecma335

AaronRobinsonMSFT committed:
Remove the statement of "undefined behavior".

AaronRobinsonMSFT committed:
Merge remote-tracking branch 'upstream/main' into update_ecma335

AaronRobinsonMSFT committed:
Updates

AaronRobinsonMSFT committed:
Updates

AaronRobinsonMSFT commented:
@agocke @jkotas or @jaredpar Can one of you take another look?

jkotas commented on docs/design/features/byreflike-generics.md:
```diff
@@ -7,22 +7,25 @@ Using ByRefLike types in Generic parameters is possible by building upon support
-- `box` &ndash; Types with ByRefLike parameters used in fields cannot be boxed.
+Providing a ByRefLike type to the `box` instruction remains invalid and `InvalidProgramException` will be thrown when detected.
+
+The `constrained.callvirt` sequence is valid if a ByRefLike type is provided. A `NotSupportedException` will be thrown at the callsite, if the target resolves to a method implemented on `object` or a default interface method.
```
```suggestion
The `constrained. callvirt` sequence is valid if a ByRefLike type is provided. A `NotSupportedException` will be thrown at the callsite, if the target resolves to a method implemented on `object` or a default interface method.
```

jkotas commented on docs/design/features/byreflike-generics.md:
```diff
@@ -7,22 +7,25 @@ Using ByRefLike types in Generic parameters is possible by building upon support
+Supporting ByRefLike types as Generic parameters will impact the following IL instructions.
 
-- `box` &ndash; Types with ByRefLike parameters used in fields cannot be boxed.
+Providing a ByRefLike type to the `box` instruction remains invalid and `InvalidProgramException` will be thrown when detected.
+
```
```suggestion
```
It would read better if it is next to the TypeLoadException case and follow the same style.

jkotas commented on docs/design/features/byreflike-generics.md:
```diff
@@ -7,22 +7,25 @@ Using ByRefLike types in Generic parameters is possible by building upon support
+
+The `constrained.callvirt` sequence is valid if a ByRefLike type is provided. A `NotSupportedException` will be thrown at the callsite, if the target resolves to a method implemented on `object` or a default interface method.
+
+Throws `TypeLoadException` when passed a ByRefLike type.
```
```suggestion
Throws `InvalidProgramException` when passed a ByRefLike type:
- `box` &ndash; ByRefLike types cannot be allocated on the heap.

Throws `TypeLoadException` when passed a ByRefLike type.
```

jkotas commented on docs/design/features/byreflike-generics.md:
```diff
@@ -110,23 +113,123 @@ throw
+
+### Option 1) Compiler helpers
+
+The following two helper functions could be introduced and would replace currently invalid `is-type` IL sequences when ByRefLike types are involved. Their behavior would broadly be defined to operate as if the ByRefLike aspect of either the `TFrom` and `TTo` is not present. An alternative approach would be consult with the Roslyn team and define the semantics of these functions to adhere to C# la ...
```
Should we agree on that we are going to do the APIs and that the API behavior will match the needs of the C# language/compiler? Or is it too early for that?

AaronRobinsonMSFT replied:
I prefer the API approach. I think @jaredpar and @AlekseyTs also are on board with the API approach. Looking for confirmation here.

jaredpar replied:
I prefer well defined runtime APIs that act as defacto IL instructions vs taking a sequence of existing IL instructions and assigning a behavior to them.

AaronRobinsonMSFT replied:
Done. Added note for .NET 10, we will adopt the API approach. Will submit new APIs and tag Roslyn team.

jkotas reviewed (approved):
LGTM as a snapshot of the current thinking. We still have work to do to decide which options to take.

AaronRobinsonMSFT committed:
Apply suggestions from code review

Co-authored-by: Jan Kotas <jkotas@microsoft.com>

AaronRobinsonMSFT commented on docs/design/features/byreflike-generics.md:
```diff
@@ -7,22 +7,26 @@ Using ByRefLike types in Generic parameters is possible by building upon support
+Throws `InvalidProgramException` when passed a ByRefLike type:
+- `box` &ndash; ByRefLike types cannot be allocated on the heap.
+
+Throws `TypeLoadException` when passed a ByRefLike type.
```
```suggestion
Throws `TypeLoadException` when passed a ByRefLike type:
```

AaronRobinsonMSFT committed:
Update docs/design/features/byreflike-generics.md

AaronRobinsonMSFT committed:
Feedback.

AaronRobinsonMSFT committed:
Updates

AaronRobinsonMSFT commented:
/ba-g System.Text.Json.Serialization.Tests.JsonElementTests.DeepEquals_TooDeepJsonDocument_ThrowsInsufficientExecutionStackException is unrelated and recently fixed - https://github.com/dotnet/runtime/issues/105425.
